zerver/actions/uploads.py

```diff
@@ -71,14 +71,23 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
     # ArchiveAttachments in the same run; prevent warnings from the
     # backing store by only removing it from there once.
     already_removed = set()
+    storage_paths = []
     for attachment in old_unclaimed_attachments:
-        delete_message_attachment(attachment.path_id)
+        storage_paths.append(attachment.path_id)
         already_removed.add(attachment.path_id)
         attachment.delete()
+        if len(storage_paths) > 1000:
+            delete_message_attachments(storage_paths)
+            storage_paths = []
     for archived_attachment in old_unclaimed_archived_attachments:
         if archived_attachment.path_id not in already_removed:
-            delete_message_attachment(archived_attachment.path_id)
+            storage_paths.append(archived_attachment.path_id)
         archived_attachment.delete()
+        if len(storage_paths) > 1000:
+            delete_message_attachments(storage_paths)
+            storage_paths = []
+    if storage_paths:
+        delete_message_attachments(storage_paths)
```

Comment:
While this is faster, it seems like it might mean a fairly in the database query for `attachment.delete` (say, due to a postgres server upgrade) can now leak hundreds of incorrectly undeleted files.

Probably the right fix would involve a more complex commit procedure, where we first mark the attachment objects as "deleted", then bulk-delete them from storage, then delete the actual `Attachment` objects from the database. 

I think such an error-handling bug existed previously at a smaller scale, though, so it might be OK to fix that in a follow-up.

Reply:
Leaking files as non-deleted in rare corner cases seems Probably Fine to me, and we do have the flag to the management command to walk the whole bucket/directory and clean out extraneous things, to recover from that.

Reply:
I've not benchmarked this, but I expect that for S3 this is a significant improvement.

Reply:
Yeah I think the right answer here is to file a follow-up issue: #30869.

Reply:
> While this is faster, it seems like it might mean a fairly in the database query for attachment.delete (say, due to a postgres server upgrade) can now leak hundreds of incorrectly undeleted files.

By "leak", do you mean that files can remain in storage after their corresponding records have been deleted from the database. ?